Preparation for transaction sequence testing. (Part 2)

jonathanknowles opened the pull request:
## Issue Number

ADP-1538

## Summary

This PR adds some more elementary building blocks for generating sequences of transactions, including:
- `selectUTxOEntries :: UTxO -> Int -> Gen ([(TxIn, TxOut)], UTxO)`
    - **Behaviour**:
        Selects up to a given number of entries at random from the given UTxO set.
    - **Intended use**:
        When generating a transaction, selecting UTxO entries to act as inputs.

- `genTokenBundlePartition :: TokenBundle -> Int -> Gen (NonEmpty TokenBundle)`
    - **Behaviour**:
        Partitions a token bundle randomly into a given number of parts.
    - **Intended use**:
        When generating a transaction, redistributing the value of selected inputs (UTxO entries) to one or more outputs.

jonathanknowles committed:
Adjust argument order of `selectMapEntries`.

This allows `selectMapEntries` to be used more conveniently in scenarios
where the number of entries is itself a generated value:

```hs
entries <- selectMapEntries m =<< chooseInt (1, 4)
```

jonathanknowles committed:
Add generator function `selectUTxOEntries` to `UTxO.Gen`.

This is just a thin wrapper around `selectMapEntries`.

jonathanknowles committed:
Add property `prop_selectMapEntries_length` to `QuickCheck.ExtraSpec`.

jonathanknowles committed:
Add generator function `chooseCoin`.

jonathanknowles committed:
Add generator function `chooseTokenQuantity`.

jonathanknowles committed:
Improve comment for function `Coin.partition`.

Emphasise that perfect proportionality may be compromised by rounding.

jonathanknowles committed:
Add function `Coin.partitionDefault`.

jonathanknowles committed:
Add function `TokenQuantity.partitionDefault`.

jonathanknowles committed:
Improve comment for function `TokenQuantity.partition`.

Emphasise that perfect proportionality may be compromised by rounding.

jonathanknowles committed:
Improve comment for function `Coin.unsafePartition`.

Emphasise that perfect proportionality may be compromised by rounding.

jonathanknowles committed:
Add generator function `genCoinPartition`.

jonathanknowles committed:
Add generator function `genTokenMapPartition`.

jonathanknowles committed:
Add generator function `genTokenBundlePartition`.

jonathanknowles committed:
Add generator function `genTokenQuantityPartition`.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/Types/UTxO/Gen.hs:
```diff
@@ -66,3 +69,15 @@ genEntryLargeRange = (,)
+-- removed.
+--
+selectUTxOEntries :: UTxO -> Int -> Gen ([(TxIn, TxOut)], UTxO)
+selectUTxOEntries = (coerce .) . selectMapEntries . unUTxO
```
Is `coerce` the same as `fromIntegral` here?

jonathanknowles replied:
> Is `coerce` the same as `fromIntegral` here?

Here `coerce` has the following type:
```hs
coerce
    :: Gen ([(TxIn, TxOut)], Map TxIn TxOut)
    -> Gen ([(TxIn, TxOut)], UTxO          )
```

It's equivalent (in this case) to `fmap (fmap UTxO)`:
```hs
selectUTxOEntries = (fmap (fmap UTxO) .) . selectMapEntries . unUTxO
```

jonathanknowles replied:
Hi @sevanspowell 

I've changed this use of `coerce` to `fmap (fmap UTxO)` in cbbb455b20d487641751d84e8add4bca343fe5b4.

sevanspowell commented on lib/core/src/Cardano/Wallet/Primitive/Types/TokenQuantity.hs:
```diff
@@ -178,8 +178,9 @@ equipartition q =
 -- | Partitions a token quantity into a number of parts, where the size of each
---   part is proportional to the size of its corresponding element in the given
---   list of weights, and the number of parts is equal to the number of weights.
+--   part is proportional (modulo rounding) to the size of its corresponding
```
With respect to rounding, `voting-tools` will assign any remainder to the last element, is that something you could consider here?

sevanspowell replied:
It makes writing laws easier and prevents a poor choice of weights compromising that `valueIn == valueOut` law.

jonathanknowles replied:
> With respect to rounding, voting-tools will assign any remainder to the last element, is that something you could consider here?

In this case, I think we want to avoid biasing any particular element in the result.

So in the (rare) case that all weights are zero, we split the target quantity as evenly as we can across the output (modulo rounding).

Note that this method _still_ has some unavoidable bias, because in the event that the target value is not evenly divisible by the number of parts, we distribute any excess to the tail of the list. (This follows from the behaviour of `equipartitionNatural` in `Numeric.Util`.

jonathanknowles committed:
Use `fmap (fmap UTxO)` instead of `coerce` in `selectUTxOEntries`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3303#discussion_r885188066
